T15 admin switch

AymenBerb opened the pull request:
Branch not finished yet, just asking for you to test it

AymenBerb committed:
add multi-path update functionality

AymenBerb committed:
remove useless reference

AymenBerb commented:
testing remains

Ali-Alkhars commented on test/widget_test.dart:
```diff
@@ -0,0 +1,30 @@
+// This is a basic Flutter widget test.
```
I'm pretty sure you didn't mean to add this file, so please delete it

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -0,0 +1,88 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/material.dart';
+
+class SwitchButton extends StatefulWidget {
```
Add a general comment explaining the use of this widget

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -0,0 +1,88 @@
+  final String itemId;
+  final String categoryItemId;
+
+  const SwitchButton(
```
catgeoryItemId is neither used nor needed. You could just remove it

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -0,0 +1,88 @@
+class _SwitchButtonState extends State<SwitchButton> {
+  bool light1 = true;
+
+  Future<List<String>> getPaths(String key) async {
```
This method is not used. Why is it here?

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -0,0 +1,88 @@
+        (querySnapshot) {
+          querySnapshot.docs.forEach((doc) {
+            FirebaseFirestore.instance
+                .collection("categoryItems")
```
For me, the switch only updates the "item" but not the "categoryItem". I get a permission denied error which usually means that the query isn't well formulated. In my branch I do this query as `.collection("categoryItems/$doc.id/items").where...` and it works, so you might want to try it out.

Or the error might be in the `FirebaseFirestore.instance.collection("categoryItems").get().then` because maybe the database rules don't allow it

Ali-Alkhars reviewed (changes requested):
Hi, nice work overall! I like how you made the switch a separate widget (I stole this idea and replicated it in my branch).

**Here are my comments other than the code comments below:**
- You need to add tests, which you already mentioned
- Catch any errors that might occur when making database calls. They do happen a lot and if not caught the user would have no clue that an error occurred and their data isn't saved. (For example, try logging in with a different user and do the switch with the hardcoded data)
- Now that all the buttons are hardcoded to the same categoryItem. Shouldn't switching one off in the UI switch all of them?

AymenBerb committed:
adopted ali's solution

AymenBerb committed:
switch functionality done

AymenBerb committed:
add test

absirgue reviewed (changes requested):
Hello sir,

Thanks for your work.

Great work in the implementation of this button. However, I feel like it is missing some tests. The fact that the multi path update works is not tested neither is the fact that the is_available is actually switched.

You can checkout Ali's branch (or mine even though it probably is less of what you want) for how to test the database.

Thanks!

Ali-Alkhars reviewed (changes requested):
Hello habibi. **The button now does change the database**, but there are other issues:

- You should make database queries catch errors. Try changing the hardcoded Id of the item that's given to the switch button to something random and click the button. It doesn't tell the user that anything wrong happened when in fact there was an error and nothing was saved.
- You still didn't make it so that all switch buttons that refer to the same itemId get switched, not just the one pressed
- Pull master where I made the temporary hardcoded data have itemIds, so now you could softcode the itemId in the switchButton.
- After pulling master, you can find a file called `firebase_functions.dart` which serves as a helper class holding all methods for manipulating Firebase. Add your methods there and use it in `admin_switch.dart`
- You should do more tests. After pulling master, go check the test file `test/services/helper_classes/firebase_functions_test.dart`. It basically mocks Firebase and gives you a test database that you could manipulate. Check how I wrote my tests to understand more (or ask me) and write actual tests for the switch availability

AymenBerb committed:
merged master

AymenBerb committed:
soft-code feature and adapt to project structure

AymenBerb committed:
add test

Ali-Alkhars commented on lib/helpers/firebase_functions.dart:
```diff
@@ -222,4 +224,53 @@ class FirebaseFunctions {
           .update({'rank': documentSnapshot.get('rank') - 1});
     }
   }
+
```
Add doc comments for your methods

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -1,6 +1,8 @@
+import 'package:seg_coursework_app/helpers/firebase_functions.dart';
 
 // ignore: slash_for_doc_comments
 /**
```
This is really unnecessary but I think it would be nice to change the comment style into /// just to stay consistent

Ali-Alkhars commented on lib/widgets/admin_switch.dart:
```diff
@@ -10,68 +12,74 @@ import 'package:flutter/material.dart';
-    final Map<String, dynamic> data =
-        documentSnapshot.data() as Map<String, dynamic>;
-    final bool? currentValue = data["is_available"];
+  Future<bool> switchBooleanValue(String itemKey) async {
```
Add a doc comment here

AymenBerb committed:
merge

Merge branch 'master' of https://github.com/AryamanDubey28/seg_coursework_app into T15_admin_switch

Ali-Alkhars reviewed (changes requested):
Hey Aymen, thank you for doing the changes. Everything works well!

**I would suggest these minor improvements though:**
- Add comments for the things I highlighted below
- Add three more unit tests: two for testing that it works the other way around (false to true). one to test that it handles errors as expected

AymenBerb committed:
add more test and comments

Ali-Alkhars reviewed (approved):
Perfect!

absirgue commented on lib/helpers/firebase_functions.dart:
```diff
@@ -73,6 +74,7 @@ class FirebaseFunctions {
     return items
         .doc(itemId)
         .update({'name': newName}).catchError((error, stackTrace) {
+      // ignore: invalid_return_type_for_catch_error
```
Can you maybe be clearer in that comment ? by saying what this error means so that someone who has no familiarity with your method or Firebase method has a sense of what's happening (you can add just "(if XXX)" the XXX being the cause that would trigger this error

absirgue reviewed (approved):
Great work! and great testing!

Congrats :)

Just a minor minor comment abt a comment that I think could be made a bit clearer so that it can be understood by everyone but this is already ready to be merged.

Thanks!
